File: Code Forces/R628/sol.cpp
#include <bits/stdc++.h>
using namespace std;

int t;
int n;
int arr[100001];
int dp[100001][100001];

// // bool check(long long x){
// // 	long double r = sqrt(x);
// // 	return ((r-floor(r)) == 0);
// // }

// int solve(int i, int cnt, long long total){
// 	if(i==n){
// 		// if(check(total) == true)
// 		// 	return cnt;
// 		return 100005;
// 	}

// 	if(dp[i][cnt] != -1) return dp[i][cnt];
// 	dp[i][cnt] = 100005;
// 	// dp[i][cnt] = min(dp[i][cnt], solve(i+1, cnt, total));
// 	// dp[i][cnt] = min(dp[i][cnt], solve(i+1, cnt+1, total*arr[i]));
// 	return dp[i][cnt];

// }

int main(){
	cin>>t;
	while(t--){
		cin>>n;
		for (int i = 0; i < n; ++i)
			cin>>arr[i];
		memset(dp, -1, sizeof dp);
		// cout<<solve(0, 0, 1)<<endl;
	}


	return 0;
}

File: Code Forces/R628/sol.py
import math
global dp
dp = [[-1]*100005]*100005

def check(x):
	r = math.sqrt(x)
	return ((r-math.floor(r))==0)

def solve(i, cnt, total):
	print(cnt, total, nums[i])
	if i>=n: 
		if(check(total)):
			return cnt
		return 100005
	if dp[i][cnt] != -1: return dp[i][cnt]

	dp[i][cnt] = 100005
	
	notake = solve(i+1, cnt, total)
	take = solve(i+1, cnt+1, total*nums[i])
	
	dp[i][cnt] = min(notake, take)
	return dp[i][cnt]

n = int(input())
nums = list(map(int, input().split()))
res = solve(0,0,1)
if(res !=100005):
	print(res)
else:
	print(-1)